### ISSUE TITLE
IllegalArgumentException | Service not registered: saschpe.android.customtabs.CustomTabsHelper

### ISSUE THREAD

**Issue by mayurwish** [saschpe/android-customtabs#5]
Fatal Exception: java.lang.RuntimeException: Unable to pause activity : java.lang.IllegalArgumentException: Service not registered: saschpe.android.customtabs.CustomTabsHelper$1@dca9585
       at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:1339)
       at android.app.ContextImpl.unbindService(ContextImpl.java:1500)
       at android.content.ContextWrapper.unbindService(ContextWrapper.java:649)
       at saschpe.android.customtabs.CustomTabsHelper.unbindCustomTabsService(SourceFile:88)
       at saschpe.android.customtabs.CustomTabsActivityLifecycleCallbacks.onActivityPaused(SourceFile:42)
       at android.app.Application.dispatchActivityPaused(Application.java:254)
       at android.app.Activity.onPause(Activity.java:1606)
       at android.support.v4.app.FragmentActivity.onPause(FragmentActivity.java:402)
       at android.app.Activity.performPause(Activity.java:6883)
       at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1455)
       at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:3889)
       at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3866)
       at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3834)
       at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3808)
       at android.app.ActivityThread.-wrap16(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1573)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6334)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)

**Comment by matecode** [comment]
I have the same issue. For me this happens when network connection is off and device is rotated.

**Comment by matecode** [comment]
As a workaround it seems to be enough not to use
```
registerActivityLifecycleCallbacks(new CustomTabsActivityLifecycleCallbacks());
```

If i don't use this in my application class, there is no error

**Comment by luongvo** [comment]
@saschpe should we catch that ex?

**Comment by saschpe** [comment]
Odd, can you guys give detail on which devices (+ Android version) this happens? The lifecycle-callbacks are optional anyway to warm-up. It's recommended but not mandatory.

**Comment by hshahdoost** [comment]
@saschpe Here is a list of android versions with this issue (among 23 users).

- 29% Android 7
- 23% - Android 6
- 18% - Android 5
- 15% - Android 4

As you can see there is nothing special about the device versions (neither the manufacturers) But there are a few things I would like to say about this error
 
1. It is very rare, (since I'm working on an app that's being run on more than 20k devices and 23 users are not a high number) still I'm very curious to know the source of this issue
2. There are devices either connected and not connected to the Internet. So I don't think it's a problem with internet connectivity.
3. From the logs I can see all users have faced this issue while working with the app, so this is not some sort of an initialization error. It seems to be somehow related to a memory leak.

Still I'm investigating the error, but would disable this feature on the next version of my app. So there won't be anymore data available on it.

**Comment by federicoboschini** [comment]
Hi everyone,
I have the same issue: very rare, seems to happen with pretty identical stacktrace while the app is in the foreground and the device is connected to the internet. The only additional info I can share is that the activity that can't be paused is styled as dialog.

**Comment by bwalsh0** [comment]
Same scenario as others described, while user is interacting with an activity - one that doesn't implement custom tabs. Here's my stacktrace:

```java
Fatal Exception: java.lang.RuntimeException: Unable to pause activity {myPkg.someAppCompatActivity}: java.lang.IllegalArgumentException: Service not registered: saschpe.android.customtabs.a@b928403
       at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:4071)
       at android.app.ActivityThread.performPauseActivity(ActivityThread.java:4026)
       at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:4795)
       at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4734)
       at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69)
       at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
       at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
       at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:55)
       at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:4784)
       at android.app.ActivityThread.access$3200(ActivityThread.java:202)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1828)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:6739)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:495)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:859)
Caused by java.lang.IllegalArgumentException: Service not registered: saschpe.android.customtabs.a@b928403
       at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:1562)
       at android.app.ContextImpl.unbindService(ContextImpl.java:1692)
       at android.content.ContextWrapper.unbindService(ContextWrapper.java:717)
       at saschpe.android.customtabs.CustomTabsHelper.unbindCustomTabsService(SourceFile:90)
       at saschpe.android.customtabs.CustomTabsActivityLifecycleCallbacks.onActivityPaused(SourceFile:42)
       at android.app.Application.dispatchActivityPaused(Application.java:248)
       at android.app.Activity.onPause(Activity.java:1733)
       at androidx.fragment.app.FragmentActivity.onPause(SourceFile:405)
       at android.app.Activity.performPause(Activity.java:7338)
       at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1466)
       at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:4061)
       at android.app.ActivityThread.performPauseActivity(ActivityThread.java:4026)
       at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:4795)
       at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4734)
       at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:69)
       at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
       at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
       at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:55)
       at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:4784)
       at android.app.ActivityThread.access$3200(ActivityThread.java:202)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1828)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:193)
       at android.app.ActivityThread.main(ActivityThread.java:6739)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:495)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:859)
```

**Comment by saschpe** [comment]
@bwalsh0 Lifecycle callbacks are run for all activities, regardless if they actually use CustomTabs or not. Pre-loading thus makes the most sense if it's used often. If your app rarely use the functionality it's probably not worth it.

Others than that, I struggle to reproduce the issue so far...

**Comment by ubuntudroid** [comment]
I have a similar stacktrace from one of our users (happened while pausing the billing library's ?? 
`ProxyBillingActivity`):

```
Fatal Exception: java.lang.RuntimeException: Unable to pause activity {my.app.id/com.android.billingclient.api.ProxyBillingActivity}: java.lang.IllegalArgumentException: Service not registered: saschpe.android.customtabs.CustomTabsHelper$a@9ad081
```

Only got it once so far, so indeed seems to happen only very rarely.

**Comment by dariush-fathie** [comment]
and i 
`E/ActivityThread: Performing pause of activity that is not resumed: {myjin.pro.ahoora.myjin.debug/myjin.pro.ahoora.myjin.ui.god.GodActivity}
    java.lang.RuntimeException: Performing pause of activity that is not resumed: {myjin.pro.ahoora.myjin.debug/myjin.pro.ahoora.myjin.ui.god.GodActivity}
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3624)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3604)
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3572)
        at android.app.ActivityThread.access$1100(ActivityThread.java:150)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1410)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:168)
        at android.app.ActivityThread.main(ActivityThread.java:5885)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:797)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:687)
V/FA: Recording user engagement, ms: 10307
V/FA: Connecting to remote service
V/FA: Activity paused, time: 505993284
E/SensorManager: uid = 10725
V/FA: Connection attempt already in progress
V/FA: Connection attempt already in progress
    
    --------- beginning of crash
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: myjin.pro.ahoora.myjin.debug, PID: 16547
    java.lang.RuntimeException: Unable to pause activity {myjin.pro.ahoora.myjin.debug/myjin.pro.ahoora.myjin.ui.god.GodActivity}: java.lang.IllegalArgumentException: Service not registered: saschpe.android.customtabs.CustomTabsHelper$bindCustomTabsService$myConnection$1@6a047a8
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3660)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3604)
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3572)
        at android.app.ActivityThread.access$1100(ActivityThread.java:150)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1410)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:168)
        at android.app.ActivityThread.main(ActivityThread.java:5885)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:797)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:687)
     Caused by: java.lang.IllegalArgumentException: Service not registered: saschpe.android.customtabs.CustomTabsHelper$bindCustomTabsService$myConnection$1@6a047a8
        at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:1100)
        at android.app.ContextImpl.unbindService(ContextImpl.java:1345)
        at android.content.ContextWrapper.unbindService(ContextWrapper.java:616)
        at saschpe.android.customtabs.CustomTabsHelper.unbindCustomTabsService(SourceFile:4)
        at saschpe.android.customtabs.CustomTabsActivityLifecycleCallbacks.onActivityPaused(SourceFile:1)
        at android.app.Application.dispatchActivityPaused(Application.java:232)
        at android.app.Activity.onPause(Activity.java:1482)
        at androidx.fragment.app.FragmentActivity.onPause(FragmentActivity.java:417)
        at myjin.pro.ahoora.myjin.ui.base.BaseActivity.onPause(BaseActivity.kt:61)
        at myjin.pro.ahoora.myjin.ui.god.GodActivity.onPause(GodActivity.kt:365)
        at android.app.Activity.performPause(Activity.java:6374)
        at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1412)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3639)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:3604) 
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:3572) 
        at android.app.ActivityThread.access$1100(ActivityThread.java:150) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1410) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:168) 
        at android.app.ActivityThread.main(ActivityThread.java:5885) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:797) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:687)`

**Comment by jonandersen** [comment]
@saschpe I started seeing this very frequently. I can reproduce it on my device 100% right now. 
I'm able to take a photo / pick from gallery but whenever the trimming/scaling view is showing the above crash happens. Not sure what more detail I can provide to be helpful, but right. I've attached a stack trace for you below:


```
  java.lang.RuntimeException: Unable to pause activity {com.public.app.development/com.publicapp.app.app.MainActivity}: java.lang.IllegalArgumentException: Service not registered: saschpe.android.customtabs.CustomTabsHelper$bindCustomTabsService$myConnection$1@8adf759
        at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:4503)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:4454)
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:4406)
        at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:228)
        at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:201)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:173)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016)
        at android.os.Handler.dispatchMessage(Handler.java:107)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7356)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
     Caused by: java.lang.IllegalArgumentException: Service not registered: saschpe.android.customtabs.CustomTabsHelper$bindCustomTabsService$myConnection$1@8adf759
        at android.app.LoadedApk.forgetServiceDispatcher(LoadedApk.java:1751)
        at android.app.ContextImpl.unbindService(ContextImpl.java:1776)
        at android.content.ContextWrapper.unbindService(ContextWrapper.java:741)
        at saschpe.android.customtabs.CustomTabsHelper.unbindCustomTabsService(SourceFile:4)
        at saschpe.android.customtabs.CustomTabsActivityLifecycleCallbacks.onActivityPaused(SourceFile:1)
        at android.app.Application.dispatchActivityPaused(Application.java:466)
        at android.app.Activity.dispatchActivityPaused(Activity.java:1306)
        at android.app.Activity.onPause(Activity.java:2228)
        at androidx.fragment.app.FragmentActivity.onPause(FragmentActivity.java:417)
        at com.publicapp.app.app.MainActivity.onPause(MainActivity.kt:109)
        at android.app.Activity.performPause(Activity.java:8001)
        at android.app.Instrumentation.callActivityOnPause(Instrumentation.java:1507)
        at android.app.ActivityThread.performPauseActivityIfNeeded(ActivityThread.java:4493)
        at android.app.ActivityThread.performPauseActivity(ActivityThread.java:4454) 
        at android.app.ActivityThread.handlePauseActivity(ActivityThread.java:4406) 
        at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:228) 
        at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:201) 
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:173) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:97) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2016) 
        at android.os.Handler.dispatchMessage(Handler.java:107) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7356) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
```

Something in here needs a check I think:

```
fun bindCustomTabsService(activity: Activity?) {
        if (client != null) {
            return
        }
        val packageName = getPackageNameToUse(activity!!) ?: return
        val myConnection = object : CustomTabsServiceConnection() {
            override fun onCustomTabsServiceConnected(
                name: ComponentName,
                newClient: CustomTabsClient
            ) {
                client = newClient
                client?.warmup(0L)
                connectionCallback?.onCustomTabsConnected()
                // Initialize a session as soon as possible.
                session
            }

            override fun onServiceDisconnected(name: ComponentName) {
                client = null
                connectionCallback?.onCustomTabsDisconnected()
            }

            override fun onBindingDied(name: ComponentName) {
                client = null
                connectionCallback?.onCustomTabsDisconnected()
            }
        }
        connection = myConnection
        CustomTabsClient.bindCustomTabsService(activity, packageName, myConnection)
    }
```

**Comment by saschpe** [comment]
Included a fix in release 3.0.2, I'll publish it this evening.
